cdap-app-templates/cdap-etl/cdap-etl-lib/src/main/java/co/cask/cdap/template/etl/transform/LogParserTransform.java

```diff
@@ -0,0 +1,203 @@
+    "event",
+    Schema.Field.of("uri", Schema.of(Schema.Type.STRING)),
+    Schema.Field.of("ip", Schema.of(Schema.Type.STRING)),
+    Schema.Field.of("browser", Schema.of(Schema.Type.STRING)),
+    Schema.Field.of("device", Schema.of(Schema.Type.STRING)),
+    Schema.Field.of("ts", Schema.of(Schema.Type.LONG))
+  );
+  private static final String LOG_FORMAT_DESCRIPTION = "Log format to parse. Currently only supports S3 and " +
+    "CLF Log format.";
+  private static final String INPUT_NAME_DESCRIPTION = "Name of the field in the input schema which encodes the " +
+    "log information. The given field must be of type String or Bytes";
+  private static final Logger LOG = LoggerFactory.getLogger(LogParserTransform.class);
+  //Regex used to parse a CLF log, each field is commented above
+  private static final Pattern CLF_LOG_PATTERN = Pattern.compile(
+    //   IP                    id    user      date          request     code     size    referrer    user agent
+    "^([\\d.]+|[:][:][\\d]) (\\S+) (\\S+) \\[([^\\]]+)\\] \"([^\"]+)\" (\\d{3}) ([-\"\\d]+) \"([^\"]+)\" \"([^\"]+)\"");
+  //Regex used to parse a S3 log, each field is commented above
+  private static final Pattern S3_LOG_PATTERN = Pattern.compile(
+    // bucket owner name   time           ip                    req   reqID operation  key    request
+    "^(\\S+) (\\S+) \\[(\\p{Print}+)\\] ([\\d.]+|[:][:][\\d]) (\\S+) (\\S+) (\\S+) (\\S+) \"([^\"]+)\" " +
+      //  HTTP stat error code  bytes sent     obj size  time    turn time  referrer         user agent
+      "(\\d{3}) (\\p{Print}+) ([-\"\\d]+) ([-\"\\d]+) ([\\d]+) ([-\"\\d]+) \"(\\p{Print}+)\" \"([^\"]+)\" " +
+      //  version id
+      "(\\p{Print}+)");
+  //Regex used to parse the request field for the URI
+  private static final Pattern REQUEST_PAGE_PATTERN = Pattern.compile("(\\S+)\\s(\\S+).*");
+  //Indices of which group request, time, ip, and user agent are in the S3 regex
+  private static final int[] S3_INDICES = {9, 3, 4, 17};
+  //Indices of which group request, time, ip, and user agent are in the CLF regex
+  private static final int[] CLF_INDICES = {5, 4, 1, 9};
+  //Number of groups matched in the S3 regex
+  private static final int S3_REGEX_LENGTH = 18;
+  //Number of groups matched in the CLF regex
+  private static final int CLF_REGEX_LENGTH = 9;
+  private static final String S3_LOG = "S3";
+  private static final String CLF_LOG = "CLF";
+  private final LogParserConfig config;
+
+  public LogParserTransform(LogParserConfig config) {
+    this.config = config;
+  }
+
+  @Override
+  public void transform(StructuredRecord input, Emitter<StructuredRecord> emitter) throws Exception {
+    String log;
+    StructuredRecord output;
+    Schema.Field inputField = input.getSchema().getField(config.inputName);
+    if (inputField == null) {
+      LOG.error("Invalid inputName, no known field matches given input.");
+      return;
+    }
+    Schema.Type inputType = inputField.getSchema().getType();
```

Comment:
what if users specify the ``Type`` as nullable? In that case, it'll be a union of ``Type.NULL`` and ``Type.STRING``/``Type.BYTES``. There is a way to get the non-nullable type from a nullable. Should we worry about this?